File: Android/app/src/androidTest/java/jp/co/riso/smartdeviceapp/view/fragment/PrintersFragmentTest.java
package jp.co.riso.smartdeviceapp.view.fragment;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import android.os.SystemClock;
import android.view.Gravity;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.test.platform.app.InstrumentationRegistry;

import org.junit.Before;
import org.junit.Test;

import jp.co.riso.smartdeviceapp.view.BaseActivityTestUtil;
import jp.co.riso.smartprint.R;

public class PrintersFragmentTest extends BaseActivityTestUtil {
    PrintersFragment mPrintersFragment = null;

    // sleep is needed because drawer calls goes through mHandler
    private void waitForDrawer() {
        SystemClock.sleep(1000);
    }

    @Before
    public void initPrintersFragment() {
        final FragmentManager fm = mainActivity.getSupportFragmentManager();

        mainActivity.runOnUiThread(() -> {
            fm.beginTransaction().add(R.id.mainLayout, new PrintersFragment()).commit();
            fm.executePendingTransactions();
        });
        InstrumentationRegistry.getInstrumentation().waitForIdleSync();

        Fragment printersFragment = fm.findFragmentById(R.id.mainLayout);
        assertTrue(printersFragment instanceof PrintersFragment);
        mPrintersFragment = (PrintersFragment) printersFragment;
    }

    @Test
    public void testNewInstance() {
        assertNotNull(mPrintersFragment);
    }

    @Test
    public void testOnClick_AddPrinter() {
        testClick(R.id.menu_id_action_add_button);
        int layoutId = R.id.mainLayout;
        if (mPrintersFragment.isTablet()) {
            waitForDrawer();
            assertTrue(mainActivity.isDrawerOpen(Gravity.RIGHT));
            layoutId = R.id.rightLayout;
        }
        final FragmentManager fm = mainActivity.getSupportFragmentManager();
        Fragment addPrinterFragment = fm.findFragmentById(layoutId);
        assertTrue(addPrinterFragment instanceof AddPrinterFragment);

        if (mPrintersFragment.isTablet()) {
            testClick(R.id.menu_id_action_add_button);
            waitForDrawer();
            assertFalse(mainActivity.isDrawerOpen(Gravity.RIGHT));
        } else {
            testClick(R.id.menu_id_back_button);
            Fragment printersFragment = fm.findFragmentById(layoutId);
            assertTrue(printersFragment instanceof PrintersFragment);
        }
    }

    @Test
    public void testOnClick_SearchPrinter() {
        testClick(R.id.menu_id_action_search_button);
        int layoutId = R.id.mainLayout;
        if (mPrintersFragment.isTablet()) {
            waitForDrawer();
            assertTrue(mainActivity.isDrawerOpen(Gravity.RIGHT));
            layoutId = R.id.rightLayout;
        }
        final FragmentManager fm = mainActivity.getSupportFragmentManager();
        Fragment searchPrintersFragment = fm.findFragmentById(layoutId);
        assertTrue(searchPrintersFragment instanceof PrinterSearchFragment);

        if (mPrintersFragment.isTablet()) {
            testClick(R.id.menu_id_action_search_button);
            waitForDrawer();
            assertFalse(mainActivity.isDrawerOpen(Gravity.RIGHT));
        } else {
            testClick(R.id.menu_id_back_button);
            Fragment printersFragment = fm.findFragmentById(layoutId);
            assertTrue(printersFragment instanceof PrintersFragment);
        }
    }

    @Test
    public void testOnClick_PrinterSearchSettings() {
        testClick(R.id.menu_id_printer_search_settings_button);
        int layoutId = R.id.mainLayout;
        if (mPrintersFragment.isTablet()) {
            waitForDrawer();
            assertTrue(mainActivity.isDrawerOpen(Gravity.RIGHT));
            layoutId = R.id.rightLayout;
        }
        final FragmentManager fm = mainActivity.getSupportFragmentManager();
        Fragment printerSearchSettingsFragment = fm.findFragmentById(layoutId);
        assertTrue(printerSearchSettingsFragment instanceof PrinterSearchSettingsFragment);

        if (mPrintersFragment.isTablet()) {
            testClick(R.id.menu_id_printer_search_settings_button);
            waitForDrawer();
            assertFalse(mainActivity.isDrawerOpen(Gravity.RIGHT));
        } else {
            testClick(R.id.menu_id_back_button);
            Fragment printersFragment = fm.findFragmentById(layoutId);
            assertTrue(printersFragment instanceof PrintersFragment);
        }
    }
}

File: Android/app/src/androidTest/java/jp/co/riso/smartdeviceapp/view/fragment/PrintersFragmentTest.kt
package jp.co.riso.smartdeviceapp.view.fragment

import android.os.SystemClock
import android.view.Gravity
import androidx.test.platform.app.InstrumentationRegistry
import jp.co.riso.smartdeviceapp.view.BaseActivityTestUtil
import jp.co.riso.smartprint.R
import org.junit.Assert
import org.junit.Before
import org.junit.Test

class PrintersFragmentTest : BaseActivityTestUtil() {
    var mPrintersFragment: PrintersFragment? = null

    // sleep is needed because drawer calls goes through mHandler
    private fun waitForDrawer() {
        SystemClock.sleep(1000)
    }

    @Before
    fun initPrintersFragment() {
        val fm = mainActivity!!.supportFragmentManager
        mainActivity!!.runOnUiThread {
            fm.beginTransaction().add(R.id.mainLayout, PrintersFragment()).commit()
            fm.executePendingTransactions()
        }
        InstrumentationRegistry.getInstrumentation().waitForIdleSync()
        val printersFragment = fm.findFragmentById(R.id.mainLayout)
        Assert.assertTrue(printersFragment is PrintersFragment)
        mPrintersFragment = printersFragment as PrintersFragment?
    }

    @Test
    fun testNewInstance() {
        Assert.assertNotNull(mPrintersFragment)
    }

    @Test
    fun testOnClick_AddPrinter() {
        testClick(R.id.menu_id_action_add_button)
        var layoutId = R.id.mainLayout
        if (mPrintersFragment!!.isTablet) {
            waitForDrawer()
            Assert.assertTrue(mainActivity!!.isDrawerOpen(Gravity.RIGHT))
            layoutId = R.id.rightLayout
        }
        val fm = mainActivity!!.supportFragmentManager
        val addPrinterFragment = fm.findFragmentById(layoutId)
        Assert.assertTrue(addPrinterFragment is AddPrinterFragment)
        if (mPrintersFragment!!.isTablet) {
            testClick(R.id.menu_id_action_add_button)
            waitForDrawer()
            Assert.assertFalse(mainActivity!!.isDrawerOpen(Gravity.RIGHT))
        } else {
            testClick(R.id.menu_id_back_button)
            val printersFragment = fm.findFragmentById(layoutId)
            Assert.assertTrue(printersFragment is PrintersFragment)
        }
    }

    @Test
    fun testOnClick_SearchPrinter() {
        testClick(R.id.menu_id_action_search_button)
        var layoutId = R.id.mainLayout
        if (mPrintersFragment!!.isTablet) {
            waitForDrawer()
            Assert.assertTrue(mainActivity!!.isDrawerOpen(Gravity.RIGHT))
            layoutId = R.id.rightLayout
        }
        val fm = mainActivity!!.supportFragmentManager
        val searchPrintersFragment = fm.findFragmentById(layoutId)
        Assert.assertTrue(searchPrintersFragment is PrinterSearchFragment)
        if (mPrintersFragment!!.isTablet) {
            testClick(R.id.menu_id_action_search_button)
            waitForDrawer()
            Assert.assertFalse(mainActivity!!.isDrawerOpen(Gravity.RIGHT))
        } else {
            testClick(R.id.menu_id_back_button)
            val printersFragment = fm.findFragmentById(layoutId)
            Assert.assertTrue(printersFragment is PrintersFragment)
        }
    }

    @Test
    fun testOnClick_PrinterSearchSettings() {
        testClick(R.id.menu_id_printer_search_settings_button)
        var layoutId = R.id.mainLayout
        if (mPrintersFragment!!.isTablet) {
            waitForDrawer()
            Assert.assertTrue(mainActivity!!.isDrawerOpen(Gravity.RIGHT))
            layoutId = R.id.rightLayout
        }
        val fm = mainActivity!!.supportFragmentManager
        val printerSearchSettingsFragment = fm.findFragmentById(layoutId)
        Assert.assertTrue(printerSearchSettingsFragment is PrinterSearchSettingsFragment)
        if (mPrintersFragment!!.isTablet) {
            testClick(R.id.menu_id_printer_search_settings_button)
            waitForDrawer()
            Assert.assertFalse(mainActivity!!.isDrawerOpen(Gravity.RIGHT))
        } else {
            testClick(R.id.menu_id_back_button)
            val printersFragment = fm.findFragmentById(layoutId)
            Assert.assertTrue(printersFragment is PrintersFragment)
        }
    }
}